sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/InstrumentationPolicy.java

```diff
@@ -164,16 +166,28 @@ private boolean isTracingEnabled(HttpPipelineCallContext context) {
         return tracer != null && tracer.isEnabled() && !((boolean) context.getData(DISABLE_TRACING_KEY).orElse(false));
     }
 
-    private final class TraceableResponse extends HttpResponse {
+    private static final class TraceableResponse extends HttpResponse {
         private final HttpResponse response;
         private final Context span;
-        private Throwable exception;
-        private String errorType;
+        private final Tracer tracer;
+        private volatile boolean ended;
```

Comment:
super nit: could look into using `AtomicIntegerReferenceUpdater` (since there isn't a boolean equivalent, and turn the `ended` checks into compare-and-set operations to do it atomically in one call.